src/rgw/driver/sfs/sqlite/sqlite_lifecycle.cc

```diff
@@ -0,0 +1,97 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t
+// vim: ts=8 sw=2 smarttab ft=cpp
+/*
+ * Ceph - scalable distributed file system
+ * SFS SAL implementation
+ *
+ * Copyright (C) 2023 SUSE LLC
+ *
+ * This is free software; you can redistribute it and/or
+ * modify it under the terms of the GNU Lesser General Public
+ * License version 2.1, as published by the Free Software
+ * Foundation. See file COPYING.
+ */
+#include "sqlite_lifecycle.h"
+
+using namespace sqlite_orm;
+namespace rgw::sal::sfs::sqlite  {
+
+SQLiteLifecycle::SQLiteLifecycle(DBConnRef _conn) : conn(_conn) { }
+
+std::optional<DBOPLCHead> SQLiteLifecycle::get_head(const std::string & oid) const {
+  auto storage = conn->get_storage();
+  auto head = storage.get_pointer<DBOPLCHead>(oid);
+  std::optional<DBOPLCHead> ret_value;
+  if (head) {
+    ret_value = *head;
+  } else {
+    // there's still no head.
+    // LC was not executed yet.
+    // create an empty entry
+    DBOPLCHead new_head{oid, "", 0};
+    storage.replace(new_head);
+    ret_value = new_head;
+  }
+  return ret_value;
+}
+
+void SQLiteLifecycle::store_head(const DBOPLCHead & head) const {
+  auto storage = conn->get_storage();
+  storage.replace(head);
+}
+
+void SQLiteLifecycle::remove_head(const std::string & oid) const {
+  auto storage = conn->get_storage();
+  storage.remove<DBOPLCHead>(oid);
+}
+
+std::optional<DBOPLEntry> SQLiteLifecycle::get_entry(const std::string & oid,
+                                            const std::string & marker) const {
+  auto storage = conn->get_storage();
+  auto db_entry = storage.get_pointer<DBOPLEntry>(oid, marker);
+  std::optional<DBOPLEntry> ret_value;
+  if (db_entry) {
+    ret_value = *db_entry;
+  }
+  return ret_value;
+}
+
+std::optional<DBOPLEntry> SQLiteLifecycle::get_next_entry(const std::string & oid,
+                                            const std::string & marker) const {
+  auto storage = conn->get_storage();
+  auto db_entries = storage.get_all<DBOPLEntry>(where(
+                                                is_equal(&DBOPLEntry::lc_index, oid) and
+                                                greater_than(&DBOPLEntry::bucket_name, marker)),
+                                                order_by(&DBOPLEntry::bucket_name).asc(),
+                                                limit(1));
+  std::optional<DBOPLEntry> ret_value;
+  // value must be unique
```

Comment:
if it *must* be unique, should we be asserting here instead of just returning a `nullopt`?

Reply:
I've changed it and it now returns an entry... because it was always returning a value anyway